File: dp/code/basic/lis-2.py
import bisect

MAXN = 100

# --8<-- [start:core]
n = 0
a = [0] * MAXN
d = [0] * MAXN
di = [0] * MAXN
pre = [0] * MAXN
res = [0] * MAXN


def dp():
    ans = 0
    for i in range(1, n + 1):
        tmp = bisect.bisect_right(d, a[i], 0, ans)
        pre[i] = di[tmp - 1] if tmp else -1
        d[tmp] = a[i]
        di[tmp] = i
        if tmp == ans:
            ans += 1

    # Construct the subsequence
    k = ans
    i = di[ans - 1]
    while k:
        res[k] = a[i]
        i = pre[i]
        k -= 1

    return ans


# --8<-- [end:core]
if __name__ == "__main__":
    t = int(input())
    while t:
        t -= 1
        n = int(input())
        a = [0] + list(map(int, input().split()))

        ans = dp()
        print(ans)
        for i in range(1, ans + 1):
            if i == ans:
                print(res[i])
            else:
                print(res[i], end=" ")

File: dp/code/basic/lis-2.cpp
#include <algorithm>
#include <cstring>
#include <iostream>

constexpr int MAXN = 100;
// --8<-- [start:core]
int n, a[MAXN], d[MAXN], di[MAXN], pre[MAXN], res[MAXN];

int dp() {
  int ans = 0;
  for (int i = 1; i <= n; ++i) {
    int tmp = std::upper_bound(d, d + ans, a[i]) - d;
    pre[i] = tmp ? di[tmp - 1] : -1;
    d[tmp] = a[i];
    di[tmp] = i;
    if (tmp == ans) ++ans;
  }
  // Construct the subsequence.
  for (int k = ans, i = di[ans - 1]; k; --k) {
    res[k] = a[i];
    i = pre[i];
  }
  return ans;
}

// --8<-- [end:core]
int main() {
  int t;
  std::cin >> t;
  while (t--) {
    std::cin >> n;
    for (int i = 1; i <= n; i++) std::cin >> a[i];
    int ans = dp();
    std::cout << ans << std::endl;
    for (int i = 1; i <= ans; ++i) {
      std::cout << res[i] << " \n"[i == ans];
    }
  }
}
